### ISSUE TITLE
Random memory bug in ParallelLinearAlgebra

### ISSUE THREAD

**Issue by dcwhite** [SCIInstitute/SCIRun#645]
This shows up sometimes in the unit tests:

```
[ RUN      ] SolveLinearSystemModuleTest.CanSolveSimple
2014-09-24 15:24:42.830  [DEBUG] Module created: SolveLinearSystem with id: SolveLinearSystem:22
2014-09-24 15:24:42.830  [DEBUG] Module algorithm initialized: SolveLinearSystem
Remark: Running algorithm Parallel cg Solver with tolerance 1e-05 and maximum iterations 0
Remark: Using preconditioner: Jacobi
libc++abi.dylib: libc++abi.dylib: libc++abi.dylib: libc++abi.dylib: libc++abi.dylib: libc++abi.dylib: terminate called throwing an exceptionlibc++abi.dylib: libc++abi.dylib: terminate called throwing an exceptionterminate called throwing an exceptionterminate called throwing an exceptionterminate called throwing an exceptionterminate called throwing an exception
terminate called throwing an exceptionAbort trap: 6
```

**Comment by dcwhite** [comment]
Update with latest error messages when it happens.

**Comment by dcwhite** [comment]
I'll add a good amount of debug logging to help reveal this one.

**Comment by dcwhite** [comment]
I think this still occurs. Need latest compiler testing.

**Comment by dcwhite** [comment]
Solved with https://github.com/SCIInstitute/SCIRun/commit/fa0996e5da9ec3656f019555b6329c09538fea4b

**Comment by dcwhite** [comment]
Reopening: this was still happening, and I think the actual root cause is now identified.

## It still reproduces

Hit today as an intermittent failure of `Algorithms_Math_Tests` on Windows:

```
C++ exception with description "Encountered an error while running parallel linear algebra"
  thrown in the test body.
[  FAILED  ] CanSolveDarrellParameterized/SolveLinearSystemTestsAlgoParameterized.CanSolveDarrellParameterized/18,
             where GetParam() = ("jacobi", "jacobi", 0)
```

Roughly **1 failure in 3 runs**, same code path and same Jacobi preconditioner as the 2014 report.

## Root cause

`ParallelLinearAlgebraSharedData` stored the per-worker result flags as:

```cpp
std::vector<bool> success_;
```

Each worker thread writes only its own slot, which reads as obviously safe:

```cpp
void ParallelLinearAlgebraBase::run_parallel(ParallelLinearAlgebraSharedData& data, int proc) const
{
  ParallelLinearAlgebra PLA(data, proc);
  data.setFlag(proc, parallel(PLA, data.inputs()));   // setFlag: success_[i] = b
}
```

But `std::vector<bool>` is the bit-packed specialization: `success_[0]`, `success_[1]`, … live in the **same underlying word**. They are not distinct memory locations, so those per-index writes are a genuine data race — each is a read-modify-write of shared state, and one worker's result can be lost or torn.

A flag that reads back false makes `start_parallel()` return false, which is exactly the "Encountered an error while running parallel linear algebra" throw in `SolveLinearSystemAlgo.cc`.

That also matches the original 2014 symptom in this issue: several threads calling `terminate` simultaneously is what you would expect when the shared flag word gets corrupted and multiple workers report failure at once.

## Why fa0996e5d didn't cover it

fa0996e5d was a correct fix for a *different* race in the same file — every thread calling `makeCompressed()` on the shared matrix:

```diff
-  mat->makeCompressed();
+  if (first())
+    mat->makeCompressed();
```

That removed one crash, but the `success_` race was two files away and survived. Reasonable call at the time; this one is invisible by inspection (no warning, compiles clean, and it works about two runs out of three).

## Fix

Use a type whose elements are distinct memory locations. Added a shared typedef next to `Parallel::RunTasks`, since this is a property of the whole `RunTasks` pattern rather than of the linear algebra:

```cpp
// Core/Thread/Parallel.h
using ParallelSuccessFlags = std::vector<char>;
```

The per-index writes are then race-free, and the results are published to the reader by the thread join at the end of `RunTasks`.

**Verified:** 12/12 consecutive passes after the change, against the prior ~1-in-3 failure rate (≈0.8% likely by chance).

## Same bug in 7 other places

The pattern had been copied across the codebase; every one of these is written from `Parallel::RunTasks` workers and had the same race:

- `Core/Algorithms/Math/ParallelAlgebra/ParallelLinearAlgebra.h` (this issue)
- `Core/Algorithms/Legacy/Fields/StreamLines/GenerateStreamLines.cc`
- `Core/Algorithms/Legacy/FiniteElements/BuildMatrix/BuildFEMatrix.cc`
- `Core/Algorithms/Legacy/FiniteElements/BuildRHS/BuildFEVolRHS.cc`
- `Core/Algorithms/Legacy/Fields/Mapping/MapFieldDataOntoElems.cc`
- `Core/Algorithms/Legacy/Fields/Mapping/MapFieldDataOntoNodes.cc`
- `Core/Algorithms/BrainStimulator/BiotSavartSolverAlgorithm.cc`
- `Core/Parser/ArrayMathInterpreter.h`

All converted to the shared typedef. This may well account for other intermittent parallel-algorithm failures.

## Possibly related

- #479 (SolveLinearSystem Jacobi does not converge with AddKnownsToLinearSystem, open)
- #1075 (SolveLinearSystem Jacobi giving bad solution, closed 2015)

Both could plausibly be this corruption surfacing as bad numbers rather than a hard failure, though neither is confirmed.

Fix is on branch `windows-unit-test-fixes-phase2`; I'll link the exact commit once the verification build and full suite finish.

**Comment by dcwhite** [comment]
Fix committed: [`f4ed6cfb8`](https://github.com/SCIInstitute/SCIRun/commit/f4ed6cfb8) on branch `windows-unit-test-fixes-phase2`.

Adds `Core::Thread::ParallelSuccessFlags` (a `std::vector<char>`) next to `Parallel::RunTasks` with the rationale documented there, and converts all eight sites that had copied the `std::vector<bool> success_` pattern.

Verification on the committed build:
- `Algorithms_Math_Tests`: **12/12** consecutive passes (previously ~1 failure in 3)
- Full unit suite: only the two known Python re-init failures remain (covered by #2551)

Regression suite is running now; I'll note the result here.

**Comment by dcwhite** [comment]
Regression suite after the fix: **308/309 passing**, no regressions. The single failure is the unrelated, separately-tracked #2567 (`editMeshBBox_trans_check` SEGFAULT in the ViewScene widget path).
